File: backend/app/schemas/scan.py
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field


class ScanType(str, Enum):
    quick = "quick"
    full = "full"


class ScanCreate(BaseModel):
    target: str = Field(
        description="IP address, hostname, or CIDR range, e.g. 127.0.0.1 or 192.168.1.0/24",
        min_length=1,
        max_length=255,
    )
    scan_type: ScanType = ScanType.quick


class ScanResultOut(BaseModel):
    id: int
    asset_id: int
    port: int
    protocol: str
    service_name: str | None
    banner: str | None

    model_config = {"from_attributes": True}


class ScanOut(BaseModel):
    id: int
    target: str
    scan_type: str
    status: str
    error_message: str | None
    started_at: datetime
    finished_at: datetime | None

    model_config = {"from_attributes": True}


class ScanDetailOut(ScanOut):
    results: list[ScanResultOut] = []

File: frontend/src/types/scan.ts
export type ScanType = "quick" | "full";
export type ScanStatus = "pending" | "running" | "completed" | "failed";

export interface ScanResult {
  id: number;
  asset_id: number;
  port: number;
  protocol: string;
  service_name: string | null;
  banner: string | null;
}

export interface Scan {
  id: number;
  target: string;
  scan_type: ScanType;
  status: ScanStatus;
  error_message: string | null;
  started_at: string;
  finished_at: string | null;
}

export interface ScanDetail extends Scan {
  results: ScanResult[];
}

export interface ScanCreatePayload {
  target: string;
  scan_type: ScanType;
}
